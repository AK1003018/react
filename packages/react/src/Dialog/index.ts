import {Dialog as DialogImpl, Header, Title, Subtitle, Body, Footer, Buttons, CloseButton} from './Dialog'
import type {DialogProps, DialogHeaderProps, DialogButtonProps, DialogWidth, DialogHeight} from './Dialog'

DialogImpl.displayName = 'Dialog'
Header.displayName = 'Dialog.Header'
Title.displayName = 'Dialog.Title'
Subtitle.displayName = 'Dialog.Subtitle'
Body.displayName = 'Dialog.Body'
Footer.displayName = 'Dialog.Footer'
Buttons.displayName = 'Dialog.Buttons'
CloseButton.displayName = 'Dialog.CloseButton'

export const Dialog = Object.assign(DialogImpl, {
  Header,
  Title,
  Subtitle,
  Body,
  Footer,
  Buttons,
  CloseButton,
})

export type {DialogProps, DialogHeaderProps, DialogButtonProps, DialogWidth, DialogHeight}
