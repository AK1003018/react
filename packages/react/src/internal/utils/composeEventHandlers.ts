type Handler<E> = (event: E) => void

export function composeEventHandlers<E>(...handlers: Array<Handler<E> | null | undefined>) {
  return function handleEvent(event: E) {
    for (const handler of handlers) {
      handler?.(event)
      if ((event as unknown as Event).defaultPrevented) {
        break
      }
    }
  }
}
