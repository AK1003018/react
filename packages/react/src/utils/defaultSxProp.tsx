import type {BetterSystemStyleObject} from '../sx'

export const defaultSxProp: Readonly<BetterSystemStyleObject> = Object.freeze<BetterSystemStyleObject>({})
