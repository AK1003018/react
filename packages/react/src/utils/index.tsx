import isNumeric from './isNumeric'
import useIsomorpyhicLayoutEffect from './useIsomorphicLayoutEffect'

export {isNumeric, useIsomorpyhicLayoutEffect}

export * from './createComponent'
export * from './defaultSxProp'
export * from './deprecate'
export * from './environment'
export * from './form-story-helpers'
export * from './getBreakpointDeclarations'
export * from './invariant'
export * from './layout'
export * from './scroll'
export * from './ssr'
export * from './theme'
export * from './use-force-update'
export * from './warning'

export type * from './types'
export type {ForwardRefComponent, OwnProps, IntrinsicElement} from './polymorphic'
