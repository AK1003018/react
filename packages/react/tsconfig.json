{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./lib",
    "typeRoots": ["../../node_modules/@types", "./node_modules/@types", "./@types"],
    "plugins": [{"name": "typescript-plugin-css-modules"}]
  },
  "include": [
    "@types/**/*.ts",
    "src/**/*.js",
    "src/**/*.ts",
    "src/**/*.tsx",
    "script/**/*.ts"
  ],
  "exclude": [
    "src/utils/story-helpers.tsx",
    "**/*.stories.tsx",
    "**/*.test.tsx",
    "**/*.figma.tsx",
    "script/**/*.ts",
    "src/__tests__/",
    "src/utils/test-*.tsx",
    "src/utils/testing.tsx",
    "e2e/**/*.ts",
    "playwright.config.ts",
  ]
}
