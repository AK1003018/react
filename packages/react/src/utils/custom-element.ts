import React from 'react'
import styled from 'styled-components'
import type {EventName} from '@lit-labs/react'
import {createComponent as create} from '@lit-labs/react'
import sx, {type SxProp} from '../sx'
import type { d } from '@storybook/channels/dist/main-c55d8855'

export interface ElementRender {
  renderShadow(): string
  shadowRootOptions?: {
    shadowrootmode?: 'open' | 'closed'
    delegatesFocus?: boolean
  }
}

function hasShadow(elementClass: object): elementClass is ElementRender {
  return 'renderShadow' in elementClass && typeof elementClass.renderShadow === 'function'
}

type EventNames = Record<string, EventName | string>
const rename = (str: string): string => str[0].toUpperCase() + str.slice(1).replace(/(-\w)/g, s => s[1].toUpperCase())
// eslint-disable-next-line @typescript-eslint/ban-types
export const createComponent = <I extends HTMLElement, E extends EventNames = {}>(
  elementClass: new () => I,
  tagName: string,
  events: E | undefined = undefined,
) => {
  const Output = Object.assign(
    styled(
      create<I, E>({
        tagName,
        elementClass,
        react: React,
        events,
      }),
    )(sx),
    {
      displayName: rename(tagName),
    },
  )

  if (hasShadow(elementClass)) {
    // Generate a key for the template
    const templateKey = Math.random().toString(36).slice(2)
    const divKey = Math.random().toString(36).slice(2)
    const dangerouslySetInnerHTML = {__html: elementClass.renderShadow()}
    // const div = React.createElement('div', {dangerouslySetInnerHTML, key: divKey})
    // const documentFragment = new DocumentFragment()
    // documentFragment.appendChild(div)
    return ({children, ...props}: React.ComponentProps<typeof Output> & SxProp) => {
      const {shadowrootmode = 'open', ...templateProps} = elementClass.shadowRootOptions || {}
      const template = React.createElement('template', {shadowrootmode, key: templateKey, dangerouslySetInnerHTML, ...templateProps})
      // @ts-ignore - Type instantiation is excessively deep and possibly infinite
      return React.createElement(Output, props, [template, ...children])
    }
  }

  return Output
}

export default createComponent
