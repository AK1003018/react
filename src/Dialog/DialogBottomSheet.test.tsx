import React from 'react'
import {fireEvent, render, waitFor} from '@testing-library/react'
import {Dialog} from './Dialog'
import MatchMediaMock from 'jest-matchmedia-mock'
import {behavesAsComponent, checkExports, renderStyles, checkStoriesForAxeViolations} from '../utils/testing'
import {mediaQueries} from '../utils/layout'
import {ANIMATION_DURATION} from './DialogBottomSheet'

let matchMedia: MatchMediaMock

describe('Dialog', () => {
  beforeEach(() => {
    matchMedia = new MatchMediaMock()
  })

  afterEach(() => {
    matchMedia.clear()
  })

  behavesAsComponent({
    Component: Dialog,
    options: {skipAs: true, skipSx: true, htmlRenderer: true},
    toRender: () => (
      <Dialog onClose={() => {}}>
        <div>Hidden when narrow</div>
      </Dialog>
    ),
  })

  checkExports('Dialog/Dialog', {
    default: undefined,
    Dialog,
  })

  it('renders `when` prop as expected', () => {
    const {container} = render(
      <Dialog onClose={() => {}}>
        <div>Hidden when narrow</div>
      </Dialog>,
    )
    expect(container).toMatchSnapshot()
  })

  // it('delays calling `onClose` when reduced motion has no preference', async () => {
  //   const onClose = jest.fn()
  //   const {getByLabelText} = render(
  //     <Dialog
  //       onClose={() => {
  //         console.log('internal onClose handler')
  //         onClose()
  //       }}
  //     >
  //       <div>Hidden when narrow</div>
  //     </Dialog>,
  //   )
  //   matchMedia.useMediaQuery('(prefers-reduced-motion)')

  //   expect(onClose).not.toHaveBeenCalled()
  //   fireEvent.click(getByLabelText('Close'))
  //   await waitFor(() => expect(onClose).toHaveBeenCalled(), {timeout: ANIMATION_DURATION + 100})
  // })
})
