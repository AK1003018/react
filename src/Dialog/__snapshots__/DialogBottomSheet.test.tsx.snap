// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog renders \`when\` prop as expected 1`] = `
@media (forced-colors:active) {

}

<div />
`;

exports[`Dialog renders consistently 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 6px;
  padding-bottom: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c6 {
  z-index: 2;
  right: 0;
  padding-right: 8px;
  position: absolute;
  pointer-events: auto;
}

.c9 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c10 {
  overflow: auto;
}

.c7 {
  border-radius: 2;
  border: 1px solid;
  border-color: var(--button-default-borderColor-rest,undefined);
  font-family: inherit;
  font-weight: semibold;
  font-size: 14px;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 32px;
  padding: 0 12px;
  gap: 8px;
  min-width: -webkit-max-content;
  min-width: -moz-max-content;
  min-width: max-content;
  -webkit-transition: 80ms cubic-bezier(0.65,0,0.35,1);
  transition: 80ms cubic-bezier(0.65,0,0.35,1);
  -webkit-transition-property: color,fill,background-color,border-color;
  transition-property: color,fill,background-color,border-color;
  color: btn.text;
  background-color: btn.bg;
  box-shadow: undefined,undefined;
}

.c7:focus:not(:disabled) {
  box-shadow: none;
  outline: 2px solid;
  outline-offset: -2px;
}

.c7:focus:not(:disabled):not(:focus-visible) {
  outline: solid 1px transparent;
}

.c7:focus-visible:not(:disabled) {
  box-shadow: none;
  outline: 2px solid;
  outline-offset: -2px;
}

.c7[href] {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c7[href]:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7:hover {
  -webkit-transition-duration: 80ms;
  transition-duration: 80ms;
}

.c7:active {
  -webkit-transition: none;
  transition: none;
}

.c7[data-inactive] {
  cursor: auto;
}

.c7:disabled {
  cursor: not-allowed;
  box-shadow: none;
  color: primer.fg.disabled;
}

.c7:disabled [data-component=ButtonCounter] {
  color: inherit;
}

.c7 [data-component=ButtonCounter] {
  font-size: 12px;
}

.c7[data-component=IconButton] {
  display: inline-grid;
  padding: unset;
  place-content: center;
  width: 32px;
  min-width: unset;
}

.c7[data-size="small"] {
  padding: 0 8px;
  height: 28px;
  gap: 4px;
  font-size: 12px;
}

.c7[data-size="small"] [data-component="text"] {
  line-height: calc(20 / 12);
}

.c7[data-size="small"] [data-component=ButtonCounter] {
  font-size: 12px;
}

.c7[data-size="small"] [data-component="buttonContent"] > :not(:last-child) {
  margin-right: 4px;
}

.c7[data-size="small"][data-component=IconButton] {
  width: 28px;
  padding: unset;
}

.c7[data-size="large"] {
  padding: 0 16px;
  height: 40px;
  gap: 8px;
}

.c7[data-size="large"] [data-component="buttonContent"] > :not(:last-child) {
  margin-right: 8px;
}

.c7[data-size="large"][data-component=IconButton] {
  width: 40px;
  padding: unset;
}

.c7[data-block="block"] {
  width: 100%;
}

.c7[data-inactive]:not([disabled]) {
  background-color: var(--button-inactive-bgColor,undefined);
  border-color: var(--button-inactive-bgColor,undefined);
  color: var(--button-inactive-fgColor,undefined);
}

.c7[data-inactive]:not([disabled]):focus-visible {
  box-shadow: none;
}

.c7 [data-component="leadingVisual"] {
  grid-area: leadingVisual;
}

.c7 [data-component="text"] {
  grid-area: text;
  line-height: calc(20/14);
  white-space: nowrap;
}

.c7 [data-component="trailingVisual"] {
  grid-area: trailingVisual;
}

.c7 [data-component="trailingAction"] {
  margin-right: -4px;
}

.c7 [data-component="buttonContent"] {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: grid;
  grid-template-areas: "leadingVisual text trailingVisual";
  grid-template-columns: min-content minmax(0,auto) min-content;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
}

.c7 [data-component="buttonContent"] > :not(:last-child) {
  margin-right: 8px;
}

.c7:hover:not([disabled]):not([data-inactive]) {
  background-color: btn.hoverBg;
  border-color: var(--button-default-borderColor-hover,undefined);
}

.c7:active:not([disabled]):not([data-inactive]) {
  background-color: btn.activeBg;
  border-color: var(--button-default-borderColor-active,undefined);
}

.c7[aria-expanded=true] {
  background-color: btn.activeBg;
  border-color: var(--button-default-borderColor-active,undefined);
}

.c7 [data-component="leadingVisual"],
.c7 [data-component="trailingVisual"],
.c7 [data-component="trailingAction"] {
  color: var(--button-color,undefined);
}

.c0 {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-animation: dialog-backdrop-appear 200ms cubic-bezier(0.33,1,0.68,1);
  animation: dialog-backdrop-appear 200ms cubic-bezier(0.33,1,0.68,1);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 296px;
  max-width: calc(100vw - 64px);
  max-height: calc(100vh - 64px);
  width: 640px;
  height: auto;
  border-radius: 12px;
  opacity: 1;
  -webkit-animation: overlay--dialog-appear 200ms cubic-bezier(0.33,1,0.68,1);
  animation: overlay--dialog-appear 200ms cubic-bezier(0.33,1,0.68,1);
}

.c2 {
  box-shadow: 0 1px 0;
  padding: 8px;
  z-index: 1;
  pointer-events: none;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
}

.c5 {
  font-size: 14px;
  font-weight: 600;
  margin: 0;
  width: calc(100% - 24px);
}

.c11 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: auto;
  padding: 16px;
}

.c8 {
  border-radius: 4px;
  background: transparent;
  border: 0;
  vertical-align: middle;
  padding: 8px;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  line-height: normal;
  box-shadow: none;
}

@media (forced-colors:active) {
  .c7:focus {
    outline: solid 1px transparent;
  }
}

<body
    style="overflow: hidden;"
  >
    <div />
    <div
      id="__primerPortalRoot__"
      style="position: absolute; top: 0px; left: 0px;"
    >
      <div
        style="position: relative; z-index: 1;"
      >
        <div
          class="c0"
        >
          <div
            aria-describedby=":r1:"
            aria-labelledby=":r0:"
            aria-modal="true"
            class="c1"
            data-focus-trap="active"
            height="auto"
            role="dialog"
            width="xlarge"
          >
            <span
              aria-hidden="true"
              class="sentinel"
              tabindex="0"
            />
            <div
              class="c2"
            >
              <div
                class="c3"
                display="flex"
              >
                <div
                  class="c4"
                  display="flex"
                >
                  <h1
                    class="c5"
                    id=":r0:"
                  >
                    Dialog
                  </h1>
                </div>
                <div
                  class="c6"
                >
                  <button
                    aria-label="Close"
                    class="c7 c8 focus-visible"
                    data-focus-visible-added=""
                    data-no-visuals="true"
                    type="button"
                  >
                    <span
                      class="c9"
                      data-component="buttonContent"
                    >
                      <span
                        data-component="text"
                      >
                        <svg
                          aria-hidden="true"
                          class=""
                          fill="currentColor"
                          focusable="false"
                          height="16"
                          role="img"
                          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                          viewBox="0 0 16 16"
                          width="16"
                        >
                          <path
                            d="M3.72 3.72a.75.75 0 0 1 1.06 0L8 6.94l3.22-3.22a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734L9.06 8l3.22 3.22a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L8 9.06l-3.22 3.22a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042L6.94 8 3.72 4.78a.75.75 0 0 1 0-1.06Z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="c10 DialogOverflowWrapper"
            >
              <div
                class="c11"
              >
                <div>
                  Hidden when narrow
                </div>
              </div>
            </div>
            <span
              aria-hidden="true"
              class="sentinel"
              tabindex="0"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": @media (forced-colors:active) {

}

<div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
